src/main/java/org/openstreetmap/atlas/geography/atlas/raw/sectioning/WaySectionProcessor.java

```diff
@@ -0,0 +1,906 @@
+     */
+    private void identifyEdgesNodesAndAreasFromLines(final WaySectionChangeSet changeSet)
+    {
+        this.rawAtlas.lines().forEach(line ->
+        {
+            if (isAtlasEdge(line))
+            {
+                final NodeOccurrenceCounter nodesForEdge = new NodeOccurrenceCounter();
+                final PolyLine polyLine = line.asPolyLine();
+
+                // Find any intersections with other edges
+                for (int index = 0; index < polyLine.size(); index++)
+                {
+                    final Location shapePoint = polyLine.get(index);
+
+                    // Based on a configurable tag filter, a shape point that doesn't intersect
+                    // an edge may also become a node
+                    if (shouldSectionAtLocation(shapePoint))
+                    {
+                        addPointToNodeList(shapePoint, nodesForEdge);
+                    }
+
+                    // TODO - Getting non-intersecting lines from the spatial query results.
+                    // So purposefully specifying "contains shapePoint". Need to resolve this!
+
+                    // If there are other edges intersecting the shape point, it becomes a node
+                    if (locationHasIntersectingLinesMatchingPredicate(shapePoint,
+                            target -> target.getIdentifier() != line.getIdentifier()
+                                    && isAtlasEdge(target)
+                                    && target.asPolyLine().contains(shapePoint)))
+                    {
+                        final Iterable<Point> pointsAtLocation = this.rawAtlas.pointsAt(shapePoint);
+                        if (Iterables.size(pointsAtLocation) > 1)
+                        {
+                            // This shouldn't happen, so let's log and section at all points. By
+                            // definition, all stacked (same location) OSM nodes will be collapsed
+                            // to a single atlas point during raw atlas creation.
+                            logger.error("Detected more than one point at {}", shapePoint);
+                        }
+                        addPointToNodeList(shapePoint, nodesForEdge);
+                    }
+                }
+
+                if (line.isClosed() && nodesForEdge.size() == 0)
+                {
+                    // Handle isolated rings by creating a node at the start and the middle
+                    // shape point, effectively splitting it in half.
+                    addPointToNodeList(polyLine.first(), nodesForEdge);
+                    addPointToNodeList(polyLine.get(polyLine.size() / 2), nodesForEdge);
+                }
+                else if (!line.isClosed())
+                {
+                    // For non-closed loops, the first and last point of the edge are always nodes.
+                    // We don't care about first/last points for rings because these can be
```

Comment:
stupid question: what's a non-closed loop?  It looks like this is guaranteeing the first and last node of a line are always added as nodes?

Reply:
This should read non-closed lines, good catch!